### ISSUE TITLE
Unexpected end of JSON input

### ISSUE THREAD

**Issue by JoshLee0915** [git-lfs/git-lfs#3679]
I am working with a small group of devs helping them get a git repo up. Their project requires them to have to move around some media files so they wanted to try to use git lfs for them. They are using a repository hosted on gitlab and I have 3 of the 4 devs running not including me, but one of them can not seem to get the lfs fetch and checkout to work. 

They keep getting an unexpected end of JSON input. It seems like this may occur with some reverse proxy or double NAT issues and they have told me that they are using a service provider that is double NATed due to the rural area they live in. 

Their environment looks like how most of the other devs environments look, including my setup. The odd thing I noticed is most of our logs for the client IP address is usually formatted like so:

> [IPv6 Addr] [IPv4 Addr]
> [IPv6 Addr] [IPv4 Addr]

But theirs looks more like this:

> [IPv6 Addr] [IPv6 Addr] [IPv6 Addr] [IPv4 Addr]

Here is the full log with some of the identifying information removed:

> git-lfs/2.7.1 (GitHub; windows amd64; go 1.11.5; git 6b7fb6e3)
> git version 2.21.0.windows.1
> 
> $ git-lfs smudge game_menu.png
> Error downloading object: game_menu.png (a56ebe6): Smudge error: Error downloading game_menu.png (a56ebe66195e6c00513cf94e8007b46095e92581324c0ff7f3c758b91911876a): batch request: unexpected end of JSON input
> 
> unexpected end of JSON input
> batch request
> github.com/git-lfs/git-lfs/errors.newWrappedError
> 	C:/Users/[User Name]/checkouts/git-lfs/errors/types.go:198
> github.com/git-lfs/git-lfs/errors.Wrap
> 	C:/Users/[User Name]/checkouts/git-lfs/errors/errors.go:74
> github.com/git-lfs/git-lfs/tq.(*tqClient).Batch
> 	C:/Users/[User Name]/checkouts/git-lfs/tq/api.go:63
> github.com/git-lfs/git-lfs/tq.Batch
> 	C:/Users/[User Name]/checkouts/git-lfs/tq/api.go:40
> github.com/git-lfs/git-lfs/tq.(*TransferQueue).enqueueAndCollectRetriesFor
> 	C:/Users/[User Name]/checkouts/git-lfs/tq/transfer_queue.go:470
> github.com/git-lfs/git-lfs/tq.(*TransferQueue).collectBatches.func1
> 	C:/Users/[User Name]/checkouts/git-lfs/tq/transfer_queue.go:391
> runtime.goexit
> 	C:/Go/src/runtime/asm_amd64.s:1333
> Error downloading game_menu.png (a56ebe66195e6c00513cf94e8007b46095e92581324c0ff7f3c758b91911876a)
> github.com/git-lfs/git-lfs/errors.newWrappedError
> 	C:/Users/[User Name]/checkouts/git-lfs/errors/types.go:198
> github.com/git-lfs/git-lfs/errors.Wrapf
> 	C:/Users/[User Name]/checkouts/git-lfs/errors/errors.go:85
> github.com/git-lfs/git-lfs/lfs.(*GitFilter).downloadFile
> 	C:/Users/[User Name]/checkouts/git-lfs/lfs/gitfilter_smudge.go:110
> github.com/git-lfs/git-lfs/lfs.(*GitFilter).Smudge
> 	C:/Users/[User Name]/checkouts/git-lfs/lfs/gitfilter_smudge.go:71
> github.com/git-lfs/git-lfs/commands.smudge
> 	C:/Users/[User Name]/checkouts/git-lfs/commands/command_smudge.go:127
> github.com/git-lfs/git-lfs/commands.smudgeCommand
> 	C:/Users/[User Name]/checkouts/git-lfs/commands/command_smudge.go:161
> github.com/spf13/cobra.(*Command).execute
> 	C:/Users/[User Name]/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:766
> github.com/spf13/cobra.(*Command).ExecuteC
> 	C:/Users/[User Name]/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852
> github.com/spf13/cobra.(*Command).Execute
> 	C:/Users/[User Name]/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
> github.com/git-lfs/git-lfs/commands.Run
> 	C:/Users/[User Name]/checkouts/git-lfs/commands/run.go:97
> main.main
> 	C:/Users/[User Name]/checkouts/git-lfs/git-lfs.go:33
> runtime.main
> 	C:/Go/src/runtime/proc.go:201
> runtime.goexit
> 	C:/Go/src/runtime/asm_amd64.s:1333
> Smudge error
> github.com/git-lfs/git-lfs/errors.newWrappedError
> 	C:/Users/[User Name]/checkouts/git-lfs/errors/types.go:198
> github.com/git-lfs/git-lfs/errors.NewSmudgeError
> 	C:/Users/[User Name]/checkouts/git-lfs/errors/types.go:284
> github.com/git-lfs/git-lfs/lfs.(*GitFilter).Smudge
> 	C:/Users/[User Name]/checkouts/git-lfs/lfs/gitfilter_smudge.go:80
> github.com/git-lfs/git-lfs/commands.smudge
> 	C:/Users/[User Name]/checkouts/git-lfs/commands/command_smudge.go:127
> github.com/git-lfs/git-lfs/commands.smudgeCommand
> 	C:/Users/[User Name]/checkouts/git-lfs/commands/command_smudge.go:161
> github.com/spf13/cobra.(*Command).execute
> 	C:/Users/[User Name]/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:766
> github.com/spf13/cobra.(*Command).ExecuteC
> 	C:/Users/[User Name]/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852
> github.com/spf13/cobra.(*Command).Execute
> 	C:/Users/[User Name]/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
> github.com/git-lfs/git-lfs/commands.Run
> 	C:/Users/[User Name]/checkouts/git-lfs/commands/run.go:97
> main.main
> 	C:/Users/[User Name]/checkouts/git-lfs/git-lfs.go:33
> runtime.main
> 	C:/Go/src/runtime/proc.go:201
> runtime.goexit
> 	C:/Go/src/runtime/asm_amd64.s:1333
> 
> Current time in UTC: 
> 2019-06-09 01:35:06
> 
> ENV:
> LocalWorkingDir=G:\[Project]
> LocalGitDir=G:\[Project]\.git
> LocalGitStorageDir=G:\[Project]\.git
> LocalMediaDir=G:\[Project]\.git\lfs\objects
> LocalReferenceDirs=
> TempDir=G:\[Project]\.git\lfs\tmp
> ConcurrentTransfers=3
> TusTransfers=false
> BasicTransfersOnly=false
> SkipDownloadErrors=false
> FetchRecentAlways=false
> FetchRecentRefsDays=7
> FetchRecentCommitsDays=0
> FetchRecentRefsIncludeRemotes=true
> PruneOffsetDays=3
> PruneVerifyRemoteAlways=false
> PruneRemoteName=origin
> LfsStorageDir=G:\[Project]\.git\lfs
> AccessDownload=none
> AccessUpload=none
> DownloadTransfers=basic
> UploadTransfers=basic
> GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
> GIT_LFS_PATH=D:\Backups\Git LFS
> GIT_SSH=C:\Program Files\PuTTY\plink.exe
> 
> Client IP addresses:
> [IPv6 Addr] [IPv6 Addr] [IPv6 Addr] [IPv4 Addr]

I am not completely sure if this is a bug and it feels more like a network issue to me, but I don't have the experience with lfs to know for certain. Has anyone else has encountered an issue like or similar to this or may have more information on how to solve it?

**Comment by bk2204** [comment]
Hey, thanks for writing in.

I think your guess that this is a network issue is probably right. We don't typically see truncated JSON as an error case. You can try running the command with `GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1` before it under Git Bash and see what output it provides, since that should dump the batch request data. That may provide fruitful information about what data you're getting and what data you're not.

You mention that you're using GitLab, but you didn't mention whether it was the cloud or self-hosted version; if the latter and you're using plain HTTP, you could try using HTTPS, which may avoid tickling something in the network if the content is the problem.

I don't think the list of IP addresses is likely an issue here. If your network is IPv4 only, then you'll only see an internal IPv6 address, but if you have an IPv6-capable network, you'll often see multiple IPv6 addresses.

**Comment by Broxzier** [comment]
My team moved our repositories, including LFS, to GitHub, and since then I ran into the same error for objects that I hadn't fetched already from our previous LFS server. With `GIT_TRACE=1` it even showed git-lfs attempting to download the files multiple times.

On my system, it was fixed by first accepting GitHub's host key. Only after running `git fetch --all` and `git lfs fetch --all` something triggered to make git ask for the host to be acknowledged. My guess is that the LFS data comes from a different server that had to be added to `known_hosts` first.
